packages/react/src/insertionFactory.ts

```diff
@@ -0,0 +1,27 @@
+import type { CSSRulesByBucket, GriffelInsertionFactory, GriffelRenderer } from '@griffel/core';
+
+import { canUseDOM } from './utils/canUseDOM';
+import { useInsertionEffect } from './useInsertionEffect';
+
+export const insertionFactory: GriffelInsertionFactory = () => {
+  const insertionCache: Record<string, boolean> = {};
+
+  return function insert(renderer: GriffelRenderer, cssRules: CSSRulesByBucket) {
+    if (useInsertionEffect) {
+      // Even if `useInsertionEffect` is available, we can't use it in SSR as it will not be executed
+      if (canUseDOM()) {
+        // eslint-disable-next-line react-hooks/rules-of-hooks
+        useInsertionEffect(() => {
+          renderer.insertCSSRules(cssRules!);
+        }, [renderer, cssRules]);
+
+        return;
+      }
+    }
+
+    if (insertionCache[renderer.id] === undefined) {
+      renderer.insertCSSRules(cssRules!);
+      insertionCache[renderer.id] = true;
+    }
+  };
+};
```

Comment:
nit: you only need one level

```suggestion
  return function insert(renderer: GriffelRenderer, cssRules: CSSRulesByBucket) {
    if (useInsertionEffect && canUseDOM) {
      // Even if `useInsertionEffect` is available, we can't use it in SSR as it will not be executed
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useInsertionEffect(() => {
        renderer.insertCSSRules(cssRules!);
      }, [renderer, cssRules]);

      return;
    }

    if (insertionCache[renderer.id] === undefined) {
      renderer.insertCSSRules(cssRules!);
      insertionCache[renderer.id] = true;
    }
  };
};
```